### PR TITLE
[RLLib] Fix OneHotPreprocessor, use gym.spaces.utils.flatten

### LINKED ISSUE #27496: [RLlib] OneHotPreprocessor one-hots MultiDiscrete Spaces incorrectly incurring information loss
### What happened + What you expected to happen

Assuming the space of `MultiDiscrete([[5 5] [5 5] [5 5] [5 5]])` :

For the observation of `[[0 0], [0 0], [0 0], [0 0]]` I'd assume the output of `[10000 10000 10000 10000 10000 10000 10000 10000 ] `, but instead we get `[10000 00000 10000 00000 10000 00000 10000 00000 ] `. For another example, for the observation of `[[2 4], [1 1], [3 3], [2 2]]` we get `[ 00101 00000 01000 00000 00010 00000 00100 00000 ]` which leads to information loss of the last two observation arrays. Instead, we should have gotten `[ 00101 00000 01000 00000 00010 00010 00100 00100 ]`, which would have preserved the second `3` and the second `2`.

I think that this is because the one-hot encoded vector for the last two arrays only encodes each number from the observation in the first set of x numbers. This is the same for all other arrays, and the last set of numbers is unused.

### Versions / Dependencies

Ray 3.0.0 dev "git_commit": "a9697722cf188ebb9b257ad7a4a6c43ece32c546"
Python 3.9
Windows 10 Enterprise 20H2

### Reproduction script

```python
from ray.rllib.models.preprocessors import OneHotPreprocessor
from gym.spaces import MultiDiscrete

space = MultiDiscrete([[5, 5], [5, 5], [5, 5], [5, 5]])
preprocess = OneHotPreprocessor(space)
dummy_vector = [[4, 1], [4, 2], [4, 3], [4, 4]]
print(preprocess.transform(dummy_vector))
```
Output `[01001 00000 00101 00000 00011 00000 00001 00000]`
vs
Expected `[00001 01000 00001 00100 00001 00010 00001 00001]`

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by olipinski**
## Why are these changes needed?

As per #27496, the current implementation of OneHotPreprocessor loses information when one-hot encoding. Using the gym implementation should avoid this issue.

## Related issue number

Closes #27496.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
nice! thanks a ton for the fix.
can you help add a quick unit test in rllib/model/tests/test_preprocess.py for this behavior?
the team is really trying to push for better quality and testing in the long terms, fix like this is perfect for adding a unit test so there won't be regression.

**Comment by olipinski** [comment]
> can you help add a quick unit test in rllib/model/tests/test_preprocess.py for this behavior?

I've added a quick test just now, both for 2D and 3D multidiscretes!

**Comment by olipinski** [comment]
Do we need to do anything else here?

**Review by gjoliver — APPROVED** [review]
thanks a ton for the fix.
I will ping @sven1977 to get this merged.

**Comment by olipinski** [comment]
Thanks for the approve!

**Comment by sven1977** [comment]
Thanks for the PR @olipinski , could you make sure the `autoregressive_action_dist` test case passes? There seems to be a connection to your changes:

https://buildkite.com/ray-project/ray-builders-pr/builds/43759#0182fa77-a584-4bef-a8ef-91f7e0ff4279

The script that fails is `rllib/examples/autoregressive_action_dist.py` run with the `--framework torch` flag.

**Comment by olipinski** [comment]
> https://buildkite.com/ray-project/ray-builders-pr/builds/43759#0182fa77-a584-4bef-a8ef-91f7e0ff4279
> 
> The script that fails is `rllib/examples/autoregressive_action_dist.py` run with the `--framework torch` flag.

This issue appears to be due to torch.nn.linear supporting only float32, which the old code used to cast every one hotted action into. I have added the cast into the preprocessor, though I'm uncertain if that should be done there, or should the dtype be specified in the environment?

**Comment by gjoliver** [comment]
I think this is fine. thanks 👍

**Comment by olipinski** [comment]
Perfect, should be all good then!

**Comment by sven1977** [comment]
Thanks again @olipinski , and @gjoliver for the review!
